src/module_library/von_caemmerer_c4_assimilation.h

```diff
@@ -0,0 +1,142 @@
+
+#ifndef VON_CAEMMERER_C4_H
+#define VON_CAEMMERER_C4_H
+
+// standard library includes
+#include <cmath>
+#include <algorithm>
+#include <array>
+#include <limits>
+
+// includes from biocro
+#include "../framework/quadratic_root.h"
+#include "../framework/constants.h"
+
+using std::array;
+
+
+
+/**
+ * @brief A object for the Michaelis Menten affinity parameters.
+ */
+class K_type {
+
+    array<double, 3> val;
+
+    public:
+        K_type(double kc, double ko, double kp) :
+         val({kc, ko, kp}) {}
+
+        double c(){return val[0];}
+        double o(){return val[1];}
+        double p(){return val[2];}
+
+};
+
+/**
+ * @brief A object for the Michaelis Menten maximum rate parameters:
+ * the maximum rubisco carboxylation rate (Vcmax), the maximum PEP carboxylation rate,
+ * the maximum PEP regeneration rate.
+ */
+class Vmax_type {
```

Comment:
I think `Vmax_type` and `K_type` could both be `struct` rather than `class`. Then you wouldn't need to write these trivial getter functions. E.g. `Vmax_type` would become 

```
struct Vmax_type {
    double rubisco_carboxylation;
    double pep_carboxylation;
    double pep_regeneration;
};
```

In BioCro, we usually use `struct` when we want to indicate that related parameters should be grouped together. There are plenty of examples, especially in `AuxBioCro.h`.

Reply:
state_map is the object type that most often should be used to pass things around. There's some instances of structs remaining  from previous code, but structs are inflexible and they aren't going to cooperate well with state_maps. Unless there's a specific reason to use a struct, I would use a state_map.

Reply:
> state_map is the object type that most often should be used to pass things around. There's some instances of structs remaining from previous code, but structs are inflexible and they aren't going to cooperate well with state_maps. Unless there's a specific reason to use a struct, I would use a state_map.

I will slightly disagree here, because there is still a pretty common scenario where a `struct` is appropriate (in my opinion). They are useful as return values of free functions that are used by multiple modules or other free functions. Details and an example are below.

It's not very straightforward to run a module from within a function, such as a free function or the `do_operation()` method of another module. So in cases where a model component needs to be used in multiple places in the C++ code, it's easier to write it as a free function.

(Side note: To run a module, the function would need to first define a state map with appropriate keys, then instantiate the module, run the module, and then retrieve the outputs from the state map. It can be done but it's a bit awkward. There is probably a way to improve this process but right now that's how it must be done. Sometimes it is worthwhile, as in the `multilayer_canopy_photosynthesis` template class, but not usually.)

An example of such a reusable code component is the FvCB model. We want it to be available on its own as a module (the `BioCro:FvCB` module), but we also need to use it in the coupled FvCB + Ball-Berry code (`c3photoC()`). So, the model is implemented via a free function called `FvCB_assim()`. The `BioCro:FvCB` module is a wrapper for this free function, and `c3photoC` also calls it.

It's typical that these free functions will need to return multiple numeric values. In that case a `struct` is used for the return value. This can also be seen in the FvCB example, where the return value of the `FvCB_assim()` function is a `struct` with values of net assimilation, Rubisco-limited assimilation, RuBP-regeneration-limited assimilation, etc.

A `state_map` would not be a good type for the input argument or return value of such a free function. If an input argument was a `state_map`, the free function would need to search through the map keys to find the values of specific elements; this is slow. If its return value was a `state_map`, then some other function would likely need to search through the map keys.

Reply:
That sounds like a reasonable use. I didn't look too closely at this code and just assumed it was inheriting and extending a module.